### PR TITLE
[SPARK-23928][SQL][WIP] Add shuffle collection function.

### PR DESCRIPTION
**Pull request by huizhilu**
## What changes were proposed in this pull request?

This PR adds a new collection function: shuffle. It generates a random permutation of the given array. This implementation uses the modern version of Fisher-Yates algorithm.

## How was this patch tested?

New tests are added to CollectionExpressionsSuite.scala and DataFrameFunctionsSuite.scala.

### CODE REVIEW

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -2268,6 +2268,21 @@ def array_sort(col):
     return Column(sc._jvm.functions.array_sort(_to_java_column(col)))
 
 
+@since(2.4)
+def shuffle(col):
+    """
+    Collection function: Generate a random permutation of the given array.
```
Comment (mn-mikke) [inline python/pyspark/sql/functions.py]: Generates / Returns

**Comment by pkuwm** [comment]
For the tests, I was trying to do this:
1. assertEqualsInorgeOrder(shuffle(originSeq), originSeq)
But spark does not have assertEqualsInorgeOrder implemented. I was thinking to check Multiset of shuffle(originSeq) and originSeq. But had trouble using Multiset for expression and seq. 
2. About the randomness, I was thinking to generate a Seq range(1, 501) and shuffle it 30 times. And it should produce at least 80% distinct permutations. Say using HashSet.add(shuffledResult). But I don't know how to implement this idea in scala and codeGen for expressions.

This is my 1st time contributing to spark and codeGen. I hope committers and contributors could help with tests, and also the shuffle function code. Thanks a lot!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -555,6 +557,100 @@ case class ArraySort(child: Expression) extends UnaryExpression with ArraySortLi
...
+@ExpressionDescription(
+  usage = "_FUNC_(array) - Returns a random permutation of the given array.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 20, 3, 5));
+       [3, 1, 5, 20]
+      > SELECT _FUNC_(array(1, 20, null, 3));
+       [20, null, 3, 1]
+  """, since = "2.4.0")
+case class Shuffle(child: Expression) extends UnaryExpression with ImplicitCastInputTypes {
```
Comment (mn-mikke) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Isn't `ExpectsInputTypes` sufficient in this case?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 3]
```diff
@@ -555,6 +557,100 @@ case class ArraySort(child: Expression) extends UnaryExpression with ArraySortLi
...
+       |for (int k = $length - 1; k >= 1; k--) {
+       |  int l = $rand.nextInt(k + 1);
+       |  $swapAssigments
+       |}
+     """.stripMargin
+  }
+
+  override protected def nullSafeEval(input: Any): Any = input match {
+    case a: ArrayData =>
+      new GenericArrayData(scala.util.
```
Comment (mn-mikke) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: line wrapping

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 4]
```diff
@@ -555,6 +557,100 @@ case class ArraySort(child: Expression) extends UnaryExpression with ArraySortLi
...
+          |for (int j = 0; j < $childName.numElements(); j++) {
+          |  ${ev.value}.update(j, ${CodeGenerator.getValue(childName, elementType, "j")});
+          |}
+       """.stripMargin
+    }
+
+    val swapAssigments = if (isPrimitiveType) {
+      val setFunc = "set" + CodeGenerator.primitiveTypeName(elementType)
+      val getCall = (index: String) => CodeGenerator.getValue(ev.value, elementType, index)
+      s"""
```
Comment (mn-mikke) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: I see strong similarities with reverse function. Would it be possible to separate common code into a new trait or class and subsequently reference it?

File: python/pyspark/sql/functions.py [hunk 5]
```diff
@@ -2268,6 +2268,21 @@ def array_sort(col):
...
+
+    :param col: name of column or expression
+
+    >>> df = spark.createDataFrame([([2, 1, 3],),([2, 1, None, 3],),([1],),([],)], ['data'])
+    >>> df.select(shuffle(df.data).alias('r')).collect()
+    [Row(r=[1, 3, 2]), Row(r=[3, None, 1, 2]), Row(r=[1]), Row(r=[])]
```
Comment (mn-mikke) [inline python/pyspark/sql/functions.py]: This will trigger a python test. Won't it fail if it's random?

Comment (pkuwm) [reply python/pyspark/sql/functions.py]: Cool. My bad. Not familiar with this. Thought they were just doc like comments... Will fix it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 4 above)
Comment (pkuwm) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Yes, I was thinking about it. But I did not want to change the code in reverse function without reviewer's comments. Will do it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 3 above)
Comment (pkuwm) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Good catch. I must hit the enter button by mistake...

**Comment by pkuwm** [comment]
@mn-mikke Can you give some comments on the tests code? I believe they need to be improved, using functions like Multiset? Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 2 above)
Comment (pkuwm) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Correct. Input is an Array. No string for input. Fixed.

**Comment by ueshin** [comment]
Thanks for your contribution!
You can refer `Uuid` and its test as I mentioned in a comment. Could you update this?
I'll revisit when you push new commits.

**Comment by ueshin** [comment]
@pkuwm Btw, if you want to implement some algorithm like "Fisher-Yates algorithm" by hand, please add a comment near the code, and a link you referred hopefully.

**Comment by ueshin** [comment]
@pkuwm Hi, any updates on this? If you have any questions, please let us know. Thanks!

**Comment by ueshin** [comment]
Okay, I'll take this over, and ping you when I submit a PR to ask a review. Thanks!

**Comment by ueshin** [comment]
@pkuwm I submitted a PR #21802 based on this. Could you take a look if you have time? Thanks!

**Comment by dongjoon-hyun** [comment]
@pkuwm . Could you close this PR since https://github.com/apache/spark/pull/21802 is merged?

**Comment by pkuwm** [comment]
Thanks for reminding, @dongjoon-hyun

**Pull request closed without merging**